File: blackjack.rb
#!/usr/bin/env ruby
# blackjack.rb
# encoding: UTF-8

require 'json'
require 'fileutils'

COLORS = {
  reset: "\e[0m",
  red: "\e[91m",
  green: "\e[92m",
  yellow: "\e[93m",
  blue: "\e[94m",
  cyan: "\e[96m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

SUITS = ['♠', '♥', '♦', '♣']
RANKS = ['2','3','4','5','6','7','8','9','10','J','Q','K','A']
VALUES = {'2'=>2,'3'=>3,'4'=>4,'5'=>5,'6'=>6,'7'=>7,'8'=>8,'9'=>9,'10'=>10,'J'=>10,'Q'=>10,'K'=>10,'A'=>11}

class Card
  attr_reader :suit, :rank, :value
  def initialize(suit, rank)
    @suit = suit
    @rank = rank
    @value = VALUES[rank]
  end
  def to_s
    color = (@suit == '♥' || @suit == '♦') ? :red : :reset
    colorize(@rank + @suit, color)
  end
end

class Deck
  attr_reader :cards
  def initialize
    @cards = []
    SUITS.each { |s| RANKS.each { |r| @cards << Card.new(s, r) } }
    shuffle
  end
  def shuffle
    @cards.shuffle!
  end
  def draw
    @cards.pop
  end
end

class Hand
  attr_accessor :cards, :bet, :done
  def initialize
    @cards = []
    @bet = 0
    @done = false
  end
  def value
    total = @cards.sum(&:value)
    aces = @cards.count { |c| c.rank == 'A' }
    while total > 21 && aces > 0
      total -= 10
      aces -= 1
    end
    total
  end
  def blackjack?
    @cards.size == 2 && value == 21
  end
  def bust?
    value > 21
  end
  def can_split?
    @cards.size == 2 && @cards[0].rank == @cards[1].rank
  end
  def split
    h = Hand.new
    h.cards << @cards.pop
    h
  end
  def to_s
    @cards.map(&:to_s).join(' ')
  end
end

class Blackjack
  attr_reader :balance, :stats_file

  def initialize(balance)
    @balance = balance
    @stats_file = File.join(Dir.home, '.blackjack_stats.json')
    load_stats
  end

  def load_stats
    if File.exist?(@stats_file)
      @stats = JSON.parse(File.read(@stats_file))
    else
      @stats = { 'wins' => 0, 'losses' => 0, 'pushes' => 0, 'blackjacks' => 0 }
    end
  end

  def save_stats
    File.write(@stats_file, JSON.pretty_generate(@stats))
  end

  def display_stats
    puts colorize("Статистика:", :bold)
    puts "  Побед: #{@stats['wins']}"
    puts "  Поражений: #{@stats['losses']}"
    puts "  Ничьих: #{@stats['pushes']}"
    puts "  Блэкджеков: #{@stats['blackjacks']}"
    puts "  Баланс: #{@balance}"
  end

  def deal_initial
    @deck = Deck.new
    @deck.shuffle
    @player_hands = [Hand.new]
    @dealer_hand = Hand.new
    @insurance = false
    @insurance_paid = 0
    @player_hands[0].cards << @deck.draw
    @dealer_hand.cards << @deck.draw
    @player_hands[0].cards << @deck.draw
    @dealer_hand.cards << @deck.draw
  end

  def show_hands(hide_dealer = true)
    puts "\nДилер:"
    if hide_dealer
      puts "  #{@dealer_hand.cards[0]}  [скрыто]"
    else
      puts "  #{@dealer_hand} (очки: #{@dealer_hand.value})"
    end
    puts "\nВаши руки:"
    @player_hands.each_with_index do |h, i|
      puts "  Рука #{i+1}: #{h} (очки: #{h.value})"
    end
  end

  def get_bet
    loop do
      print "Ваша ставка (баланс: #{@balance}): "
      input = gets.chomp
      bet = input.to_i
      if bet > 0 && bet <= @balance
        @bet = bet
        return
      end
      puts colorize("Неверная ставка.", :red)
    end
  end

  def play_hand(idx)
    hand = @player_hands[idx]
    while !hand.done && !hand.bust? && hand.value < 21
      show_hands
      puts "\nРука #{idx+1}: #{hand} (очки: #{hand.value})"
      print "Действие (h=hit, s=stand, d=double, i=insurance, sp=split, q=quit): "
      action = gets.chomp.strip.downcase
      case action
      when 'q'
        puts colorize("Выход.", :yellow)
        save_stats
        exit
      when 'h'
        hand.cards << @deck.draw
        if hand.bust?
          puts colorize("Перебор!", :red)
          hand.done = true
        end
      when 's'
        hand.done = true
      when 'd'
        if hand.cards.size == 2 && @balance >= @bet
          @bet *= 2
          hand.cards << @deck.draw
          hand.done = true
          puts colorize("Перебор!", :red) if hand.bust?
        else
          puts colorize("Удвоение недоступно.", :yellow)
        end
      when 'i'
        if @dealer_hand.cards[0].rank == 'A' && !@insurance
          @insurance = true
          @insurance_paid = @bet / 2
          @balance -= @insurance_paid
          puts colorize("Страховка активирована.", :cyan)
        else
          puts colorize("Страховка недоступна.", :yellow)
        end
      when 'sp'
        if hand.can_split?
          new_hand = hand.split
          @player_hands << new_hand
          hand.cards << @deck.draw
          new_hand.cards << @deck.draw
          puts colorize("Руки разделены.", :cyan)
          if @balance < @bet
            puts colorize("Недостаточно средств для сплита.", :red)
            @player_hands.pop
          else
            @balance -= @bet
            @bet *= 2
          end
        else
          puts colorize("Сплит недоступен.", :yellow)
        end
      else
        puts colorize("Неизвестная команда.", :red)
      end
    end
  end

  def dealer_play
    while @dealer_hand.value < 17
      @dealer_hand.cards << @deck.draw
    end
  end

  def resolve
    show_hands(false)
    dealer_val = @dealer_hand.value
    @player_hands.each do |hand|
      player_val = hand.value
      if hand.bust?
        @balance -= @bet
        @stats['losses'] += 1
        puts colorize("Поражение (перебор): рука проиграла #{@bet}", :red)
      elsif hand.blackjack? && !@dealer_hand.blackjack?
        win = (@bet * 1.5).to_i
        @balance += win
        @stats['blackjacks'] += 1
        @stats['wins'] += 1
        puts colorize("Блэкджек! Выигрыш #{win}", :green)
      elsif dealer_val > 21
        @balance += @bet
        @stats['wins'] += 1
        puts colorize("Дилер перебрал, выигрыш #{@bet}", :green)
      elsif player_val > dealer_val
        @balance += @bet
        @stats['wins'] += 1
        puts colorize("Выигрыш #{@bet}", :green)
      elsif player_val == dealer_val
        @stats['pushes'] += 1
        puts colorize("Ничья, ставка возвращена.", :yellow)
      else
        @balance -= @bet
        @stats['losses'] += 1
        puts colorize("Поражение, потеряно #{@bet}", :red)
      end
    end
    if @insurance
      if @dealer_hand.blackjack?
        @balance += @insurance_paid * 2
        puts colorize("Страховка выиграла!", :cyan)
      else
        puts colorize("Страховка проиграла.", :yellow)
      end
    end
    save_stats
  end

  def play
    puts colorize("🃏 Добро пожаловать в Блэкджек!", :bold)
    puts "Ваш баланс: #{@balance}"
    while @balance > 0
      get_bet
      deal_initial
      if @player_hands[0].blackjack?
        show_hands
        if @dealer_hand.blackjack?
          puts colorize("Ничья – у обоих блэкджек!", :yellow)
          @stats['pushes'] += 1
          save_stats
          next
        else
          win = (@bet * 1.5).to_i
          @balance += win
          @stats['blackjacks'] += 1
          @stats['wins'] += 1
          puts colorize("Блэкджек! Выигрыш #{win}", :green)
          save_stats
          next
        end
      end
      if @dealer_hand.cards[0].rank == 'A'
        print "У дилера туз. Хотите страховку? (y/n): "
        ans = gets.chomp.strip.downcase
        if ans == 'y'
          @insurance = true
          @insurance_paid = @bet / 2
          @balance -= @insurance_paid
          puts colorize("Страховка активирована.", :cyan)
        end
      end
      @player_hands.each_with_index { |_, i| play_hand(i) }
      dealer_play
      resolve
      puts colorize("Баланс: #{@balance}", :bold)
      if @balance <= 0
        puts colorize("Вы проиграли все деньги!", :red)
        break
      end
      print "Продолжить игру? (y/n): "
      cont = gets.chomp.strip.downcase
      break if cont != 'y'
    end
    display_stats
    save_stats
  end
end

def main
  balance = 1000
  reset = false
  i = 1
  while i < ARGV.size
    arg = ARGV[i]
    case arg
    when '-r', '--reset' then reset = true
    when '-s', '--start'
      balance = ARGV[i+1].to_i if i+1 < ARGV.size
      i += 1
    when '-h', '--help'
      puts "Usage: ruby blackjack.rb [-s start_balance] [-r]"
      return
    end
    i += 1
  end
  if reset
    f = File.join(Dir.home, '.blackjack_stats.json')
    File.delete(f) if File.exist?(f)
    puts "Статистика сброшена."
    return
  end
  game = Blackjack.new(balance)
  game.play
end

main if __FILE__ == $0

File: blackjack.cs
// blackjack.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Linq;

class BlackjackGame
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "red" => "\x1b[91m",
            "green" => "\x1b[92m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            "cyan" => "\x1b[96m",
            "bold" => "\x1b[1m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static string[] SUITS = {"♠","♥","♦","♣"};
    static string[] RANKS = {"2","3","4","5","6","7","8","9","10","J","Q","K","A"};
    static Dictionary<string, int> VALUES = new Dictionary<string, int> {
        {"2",2},{"3",3},{"4",4},{"5",5},{"6",6},{"7",7},{"8",8},{"9",9},{"10",10},
        {"J",10},{"Q",10},{"K",10},{"A",11}
    };

    class Card
    {
        public string Suit { get; }
        public string Rank { get; }
        public int Value { get; }
        public Card(string suit, string rank)
        {
            Suit = suit; Rank = rank; Value = VALUES[rank];
        }
        public override string ToString()
        {
            string col = (Suit == "♥" || Suit == "♦") ? "red" : "reset";
            return Colorize(Rank + Suit, col);
        }
    }

    class Deck
    {
        public List<Card> Cards { get; set; }
        public Deck()
        {
            Cards = new List<Card>();
            foreach (var s in SUITS)
                foreach (var r in RANKS)
                    Cards.Add(new Card(s, r));
            Shuffle();
        }
        public void Shuffle()
        {
            Random rnd = new Random();
            for (int i = Cards.Count - 1; i > 0; i--)
            {
                int j = rnd.Next(i + 1);
                var tmp = Cards[i];
                Cards[i] = Cards[j];
                Cards[j] = tmp;
            }
        }
        public Card Draw()
        {
            if (Cards.Count == 0) return null;
            var c = Cards[Cards.Count - 1];
            Cards.RemoveAt(Cards.Count - 1);
            return c;
        }
    }

    class Hand
    {
        public List<Card> Cards { get; set; } = new List<Card>();
        public int Bet { get; set; }
        public bool Done { get; set; }
        public int Value()
        {
            int total = Cards.Sum(c => c.Value);
            int aces = Cards.Count(c => c.Rank == "A");
            while (total > 21 && aces > 0) { total -= 10; aces--; }
            return total;
        }
        public bool IsBlackjack() => Cards.Count == 2 && Value() == 21;
        public bool IsBust() => Value() > 21;
        public bool CanSplit() => Cards.Count == 2 && Cards[0].Rank == Cards[1].Rank;
        public Hand Split()
        {
            Hand h = new Hand();
            h.Cards.Add(Cards[Cards.Count - 1]);
            Cards.RemoveAt(Cards.Count - 1);
            return h;
        }
        public override string ToString() => string.Join(" ", Cards.Select(c => c.ToString()));
    }

    class Stats
    {
        public int wins { get; set; }
        public int losses { get; set; }
        public int pushes { get; set; }
        public int blackjacks { get; set; }
    }

    class Blackjack
    {
        public int Balance { get; set; }
        public int Bet { get; set; }
        public Deck Deck { get; set; }
        public List<Hand> PlayerHands { get; set; }
        public Hand DealerHand { get; set; }
        public bool Insurance { get; set; }
        public int InsurancePaid { get; set; }
        public Stats Stats { get; set; }
        public string StatsFile { get; }

        public Blackjack(int balance)
        {
            Balance = balance;
            StatsFile = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".blackjack_stats.json");
            LoadStats();
        }

        void LoadStats()
        {
            if (File.Exists(StatsFile))
            {
                string json = File.ReadAllText(StatsFile);
                Stats = JsonSerializer.Deserialize<Stats>(json) ?? new Stats();
            }
            else Stats = new Stats();
        }

        void SaveStats()
        {
            string json = JsonSerializer.Serialize(Stats);
            File.WriteAllText(StatsFile, json);
        }

        void DisplayStats()
        {
            Console.WriteLine(Colorize("Статистика:", "bold"));
            Console.WriteLine($"  Побед: {Stats.wins}");
            Console.WriteLine($"  Поражений: {Stats.losses}");
            Console.WriteLine($"  Ничьих: {Stats.pushes}");
            Console.WriteLine($"  Блэкджеков: {Stats.blackjacks}");
            Console.WriteLine($"  Баланс: {Balance}");
        }

        void DealInitial()
        {
            Deck = new Deck();
            Deck.Shuffle();
            PlayerHands = new List<Hand> { new Hand() };
            DealerHand = new Hand();
            Insurance = false;
            InsurancePaid = 0;
            PlayerHands[0].Cards.Add(Deck.Draw());
            DealerHand.Cards.Add(Deck.Draw());
            PlayerHands[0].Cards.Add(Deck.Draw());
            DealerHand.Cards.Add(Deck.Draw());
        }

        void ShowHands(bool hideDealer = true)
        {
            Console.WriteLine("\nДилер:");
            if (hideDealer)
                Console.WriteLine($"  {DealerHand.Cards[0].ToString()}  [скрыто]");
            else
                Console.WriteLine($"  {DealerHand} (очки: {DealerHand.Value()})");
            Console.WriteLine("\nВаши руки:");
            for (int i=0; i<PlayerHands.Count; i++)
                Console.WriteLine($"  Рука {i+1}: {PlayerHands[i]} (очки: {PlayerHands[i].Value()})");
        }

        void GetBet()
        {
            while (true)
            {
                Console.Write($"Ваша ставка (баланс: {Balance}): ");
                string input = Console.ReadLine();
                if (int.TryParse(input, out int bet) && bet > 0 && bet <= Balance)
                {
                    Bet = bet;
                    return;
                }
                Console.WriteLine(Colorize("Неверная ставка.", "red"));
            }
        }

        void PlayHand(int idx)
        {
            var hand = PlayerHands[idx];
            while (!hand.Done && !hand.IsBust() && hand.Value() < 21)
            {
                ShowHands();
                Console.WriteLine($"\nРука {idx+1}: {hand} (очки: {hand.Value()})");
                Console.Write("Действие (h=hit, s=stand, d=double, i=insurance, sp=split, q=quit): ");
                string action = Console.ReadLine().Trim().ToLower();
                switch (action)
                {
                    case "q":
                        Console.WriteLine(Colorize("Выход.", "yellow"));
                        SaveStats();
                        Environment.Exit(0);
                        break;
                    case "h":
                        hand.Cards.Add(Deck.Draw());
                        if (hand.IsBust()) { Console.WriteLine(Colorize("Перебор!", "red")); hand.Done = true; }
                        break;
                    case "s":
                        hand.Done = true;
                        break;
                    case "d":
                        if (hand.Cards.Count == 2 && Balance >= Bet)
                        {
                            Bet *= 2;
                            hand.Cards.Add(Deck.Draw());
                            hand.Done = true;
                            if (hand.IsBust()) Console.WriteLine(Colorize("Перебор!", "red"));
                        }
                        else Console.WriteLine(Colorize("Удвоение недоступно.", "yellow"));
                        break;
                    case "i":
                        if (DealerHand.Cards[0].Rank == "A" && !Insurance)
                        {
                            Insurance = true;
                            InsurancePaid = Bet / 2;
                            Balance -= InsurancePaid;
                            Console.WriteLine(Colorize("Страховка активирована.", "cyan"));
                        }
                        else Console.WriteLine(Colorize("Страховка недоступна.", "yellow"));
                        break;
                    case "sp":
                        if (hand.CanSplit())
                        {
                            var newHand = hand.Split();
                            PlayerHands.Add(newHand);
                            hand.Cards.Add(Deck.Draw());
                            newHand.Cards.Add(Deck.Draw());
                            Console.WriteLine(Colorize("Руки разделены.", "cyan"));
                            if (Balance < Bet)
                            {
                                Console.WriteLine(Colorize("Недостаточно средств для сплита.", "red"));
                                PlayerHands.RemoveAt(PlayerHands.Count - 1);
                            }
                            else
                            {
                                Balance -= Bet;
                                Bet *= 2;
                            }
                        }
                        else Console.WriteLine(Colorize("Сплит недоступен.", "yellow"));
                        break;
                    default:
                        Console.WriteLine(Colorize("Неизвестная команда.", "red"));
                        break;
                }
            }
        }

        void DealerPlay()
        {
            while (DealerHand.Value() < 17)
                DealerHand.Cards.Add(Deck.Draw());
        }

        void Resolve()
        {
            ShowHands(false);
            int dealerVal = DealerHand.Value();
            foreach (var hand in PlayerHands)
            {
                int playerVal = hand.Value();
                if (hand.IsBust())
                {
                    Balance -= Bet;
                    Stats.losses++;
                    Console.WriteLine(Colorize($"Поражение (перебор): рука проиграла {Bet}", "red"));
                }
                else if (hand.IsBlackjack() && !DealerHand.IsBlackjack())
                {
                    int win = (int)(Bet * 1.5);
                    Balance += win;
                    Stats.blackjacks++;
                    Stats.wins++;
                    Console.WriteLine(Colorize($"Блэкджек! Выигрыш {win}", "green"));
                }
                else if (dealerVal > 21)
                {
                    Balance += Bet;
                    Stats.wins++;
                    Console.WriteLine(Colorize($"Дилер перебрал, выигрыш {Bet}", "green"));
                }
                else if (playerVal > dealerVal)
                {
                    Balance += Bet;
                    Stats.wins++;
                    Console.WriteLine(Colorize($"Выигрыш {Bet}", "green"));
                }
                else if (playerVal == dealerVal)
                {
                    Stats.pushes++;
                    Console.WriteLine(Colorize("Ничья, ставка возвращена.", "yellow"));
                }
                else
                {
                    Balance -= Bet;
                    Stats.losses++;
                    Console.WriteLine(Colorize($"Поражение, потеряно {Bet}", "red"));
                }
            }
            if (Insurance)
            {
                if (DealerHand.IsBlackjack())
                {
                    Balance += InsurancePaid * 2;
                    Console.WriteLine(Colorize("Страховка выиграла!", "cyan"));
                }
                else Console.WriteLine(Colorize("Страховка проиграла.", "yellow"));
            }
            SaveStats();
        }

        public void Play()
        {
            Console.WriteLine(Colorize("🃏 Добро пожаловать в Блэкджек!", "bold"));
            Console.WriteLine($"Ваш баланс: {Balance}");
            while (Balance > 0)
            {
                GetBet();
                DealInitial();
                if (PlayerHands[0].IsBlackjack())
                {
                    ShowHands();
                    if (DealerHand.IsBlackjack())
                    {
                        Console.WriteLine(Colorize("Ничья – у обоих блэкджек!", "yellow"));
                        Stats.pushes++;
                        SaveStats();
                        continue;
                    }
                    else
                    {
                        int win = (int)(Bet * 1.5);
                        Balance += win;
                        Stats.blackjacks++;
                        Stats.wins++;
                        Console.WriteLine(Colorize($"Блэкджек! Выигрыш {win}", "green"));
                        SaveStats();
                        continue;
                    }
                }
                if (DealerHand.Cards[0].Rank == "A")
                {
                    Console.Write("У дилера туз. Хотите страховку? (y/n): ");
                    string ans = Console.ReadLine().Trim().ToLower();
                    if (ans == "y")
                    {
                        Insurance = true;
                        InsurancePaid = Bet / 2;
                        Balance -= InsurancePaid;
                        Console.WriteLine(Colorize("Страховка активирована.", "cyan"));
                    }
                }
                for (int i=0; i<PlayerHands.Count; i++)
                    PlayHand(i);
                DealerPlay();
                Resolve();
                Console.WriteLine(Colorize($"Баланс: {Balance}", "bold"));
                if (Balance <= 0)
                {
                    Console.WriteLine(Colorize("Вы проиграли все деньги!", "red"));
                    break;
                }
                Console.Write("Продолжить игру? (y/n): ");
                string cont = Console.ReadLine().Trim().ToLower();
                if (cont != "y") break;
            }
            DisplayStats();
            SaveStats();
        }
    }

    static void Main(string[] args)
    {
        int balance = 1000;
        bool reset = false;
        for (int i=0; i<args.Length; i++)
        {
            if (args[i] == "-r" || args[i] == "--reset") reset = true;
            else if ((args[i] == "-s" || args[i] == "--start") && i+1 < args.Length)
                balance = int.Parse(args[++i]);
            else if (args[i] == "-h" || args[i] == "--help")
            {
                Console.WriteLine("Usage: blackjack [-s start_balance] [-r]");
                return;
            }
        }
        if (reset)
        {
            string f = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".blackjack_stats.json");
            if (File.Exists(f)) File.Delete(f);
            Console.WriteLine("Статистика сброшена.");
            return;
        }
        Blackjack game = new Blackjack(balance);
        game.Play();
    }
}
